Bottom: d90d38e79fd9560c94785d2a95b95962d3fcbec3
Top:    5a3057114e749775a8f90e29e6f96c3f9549701e
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-05-14 00:46:27 +0100

GTK/GL: Add facility to set an alpha multiplier for the current rendering


---

diff --git a/src/hid/gtk/gtkhid-gl.c b/src/hid/gtk/gtkhid-gl.c
index cfd41ff..f5c4dc9 100644
--- a/src/hid/gtk/gtkhid-gl.c
+++ b/src/hid/gtk/gtkhid-gl.c
@@ -46,6 +46,7 @@ typedef struct render_priv {
   bool in_context;
   int subcomposite_stencil_bit;
   char *current_colorname;
+  double current_alpha_mult;
 } render_priv;
 
 
@@ -54,6 +55,7 @@ typedef struct hid_gc_struct
   HID *me_pointer;
 
   const char *colorname;
+  double alpha_mult;
   gint width;
   gint cap, join;
   gchar xor;
@@ -172,6 +174,7 @@ ghid_make_gc (void)
   rv = g_new0 (hid_gc_struct, 1);
   rv->me_pointer = &ghid_hid;
   rv->colorname = Settings.BackgroundColor;
+  rv->alpha_mult = 1.0;
   return rv;
 }
 
@@ -358,11 +361,13 @@ set_gl_color_for_gc (hidGC gc)
   a = 1.0;
 
   if (priv->current_colorname != NULL &&
-      strcmp (priv->current_colorname, gc->colorname) == 0)
+      strcmp (priv->current_colorname, gc->colorname) == 0 &&
+      priv->current_alpha_mult == gc->alpha_mult)
     return;
 
   free (priv->current_colorname);
   priv->current_colorname = strdup (gc->colorname);
+  priv->current_alpha_mult = gc->alpha_mult;
 
   if (gport->colormap == NULL)
     gport->colormap = gtk_widget_get_colormap (gport->top_window);
@@ -423,6 +428,7 @@ set_gl_color_for_gc (hidGC gc)
     }
   if (1) {
     double maxi, mult;
+    alpha_mult *= gc->alpha_mult;
     if (priv->trans_lines)
       a = a * alpha_mult;
     maxi = r;
@@ -451,6 +457,13 @@ ghid_set_color (hidGC gc, const char *name)
 }
 
 void
+ghid_set_alpha_mult (hidGC gc, double alpha_mult)
+{
+  gc->alpha_mult = alpha_mult;
+  set_gl_color_for_gc (gc);
+}
+
+void
 ghid_set_line_cap (hidGC gc, EndCapStyle style)
 {
   gc->cap = style;
